ymca/src/main/java/endpoints/LoggingEndpoint.java

```diff
@@ -67,65 +86,142 @@ public void onClose(final Session session, final CloseReason closeReason)
 	{
-		final JSONObject jsonMessage = new JSONObject();
-		jsonMessage.put("level", level.getName());
-		jsonMessage.put("message", message);
-
+		final String sessionId = logMessage.getSessionId();
+		
 		// If sessionID is not specified, notify all endpoints
 		if (null == sessionId)
 		{
-			for (final Set<LoggingEndpoint> set : LoggingEndpoint.endpoints.values())
+			for (final Set<LoggingEndpoint> endpoints : LoggingEndpoint.endpoints.values())
 			{
-				for (final LoggingEndpoint endpoint : set)
-				{
-					endpoint.session.getAsyncRemote().sendText(jsonMessage.toString());
-				}
+				logToClient(endpoints, logMessage);
+			}
+
+			// Queue message for all sessions
+			if (queue)
+			{
+				queueMessageToAll(logMessage);
 			}
 		}
 		else
 		{
 			final Set<LoggingEndpoint> sessionEndpoints = LoggingEndpoint.endpoints.get(sessionId);
 			if (null != sessionEndpoints)
 			{
-				for (final LoggingEndpoint endpoint : sessionEndpoints)
+				logToClient(sessionEndpoints, logMessage);
+
+				if (queue)
 				{
-					endpoint.session.getAsyncRemote().sendText(jsonMessage.toString());
+					queueMessage(logMessage);
 				}
 			}
 			else
 			{
 				System.err.println(String.format("No LoggingEndpoints regestered for session '%s'", sessionId));
-				System.err.println(level.getName() + ": " + message);
+				System.err.println(logMessage);
 			}
 		}
 	}
+
+	private static void queueMessageToAll(final LogMessage logMessage)
+	{
+		for (final SortedSet<LogMessage> messageSet : LoggingEndpoint.messages.values())
+		{
+			messageSet.add(logMessage);
+		}
+	}
+	
+
+	/**
+	 * Adds a message to the queue of messages for that session ID
+	 * 
+	 * @param sessionId
+	 *            The HTTP session ID
+	 * @param logMessage
+	 *            The message to queue
+	 */
+	private static void queueMessage(final LogMessage logMessage)
+	{
+		SortedSet<LogMessage> messageQueue = LoggingEndpoint.messages.get(logMessage.getSessionId());
+
+		if (null == messageQueue)
+		{
+			messageQueue = new TreeSet<>();
+		}
+
+		messageQueue.add(logMessage);
+		LoggingEndpoint.messages.put(logMessage.getSessionId(), messageQueue);
+	}
+
+	/**
+	 * Returns the queue of existing messages for a given session
+	 * 
+	 * @param sessionId
+	 *            The session ID for the session messages to be retrieved
+	 * @return The queue of messages
+	 */
+	public static SortedSet<LogMessage> getMessages(final String sessionId)
+	{
+		return LoggingEndpoint.messages.get(sessionId);
+	}
+
+	/**
+	 * Logs the JSON message to every provided endpoint
+	 * 
+	 * @param endpoints
+	 *            The endpoints to send the message to
+	 * @param message
+	 *            The message
+	 */
+	private static void logToClient(final Set<LoggingEndpoint> endpoints, final LogMessage message)
+	{
+		if (null != endpoints)
+		{
+			for (final LoggingEndpoint endpoint : endpoints)
+			{
+				try
+				{
+					endpoint.session.getBasicRemote().sendText(message.toJSONString());
```

Comment:
Makes more sense to me to only call .toJSONString in one place, rather than every time you want to call _logToClient_